x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/BasicQueryClient.java

```diff
@@ -67,37 +68,35 @@ protected void search(SearchRequest search, ActionListener<SearchResponse> liste
         client.search(search, listener);
     }
 
+    protected void search(MultiSearchRequest search, ActionListener<MultiSearchResponse> listener) {
+        client.multiSearch(search, listener);
+    }
+
     @Override
     public void fetchHits(Iterable<List<HitReference>> refs, ActionListener<List<List<SearchHit>>> listener) {
-        IdsQueryBuilder idsQuery = idsQuery();
-
         int innerListSize = 0;
-        Set<String> indices = new HashSet<>();
 
-        // associate each reference with its own
+        Map<String, IdsQueryBuilder> queries = new HashMap<>();
+
+        // associate each reference with its positions inside the matrix
         final Map<HitReference, List<Integer>> referenceToPosition = new HashMap<>();
         int counter = 0;
 
         for (List<HitReference> list : refs) {
             innerListSize = list.size();
             for (HitReference ref : list) {
-                idsQuery.addIds(ref.id());
-                indices.add(ref.index());
-                // remember the reference position
+                // keep ids per index
+                IdsQueryBuilder query = queries.computeIfAbsent(ref.index(), v -> idsQuery());
```

Comment:
You also need to filter by index too since the query will hit all the indices inside the PIT so you might have duplicates ? So two filter clauses are needed, one for the ids and one to match the right index.